# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsIGlobalObject.cpp
# Commit: a8e623b228ae
# Full Hash: a8e623b228aeb04c741bebeeaea4f2524b2ddcc7
# Author: Tom Ritter <tom@mozilla.com>
# Date: 2022-11-29 21:47:49
# Regressor Bug: 1778510
# File Overlap Count: 1
# Description:
#   Bug 1778510: Add RTPCallerType to GlobalObject r=asuth
#   
#   This centralizes the logic in one place.
#   
#   In order to do this, we will need to check the principal
# ==============================================================================

diff -r 32fdf6f014f7 -r a8e623b228ae dom/base/nsIGlobalObject.cpp
--- a/dom/base/nsIGlobalObject.cpp	Tue Nov 29 13:34:14 2022 +0000
+++ b/dom/base/nsIGlobalObject.cpp	Tue Nov 29 13:34:15 2022 +0000
@@ -5,6 +5,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "nsIGlobalObject.h"
+#include "mozilla/BasePrincipal.h"
 #include "mozilla/CycleCollectedJSContext.h"
 #include "mozilla/Result.h"
 #include "mozilla/StorageAccess.h"
@@ -68,7 +69,7 @@
   MOZ_DIAGNOSTIC_ASSERT(mEventTargetObjects.isEmpty());
 }
 
-nsIPrincipal* nsIGlobalObject::PrincipalOrNull() {
+nsIPrincipal* nsIGlobalObject::PrincipalOrNull() const {
   if (!NS_IsMainThread()) {
     return nullptr;
   }
@@ -396,3 +397,26 @@
 
   return mozilla::ipc::StorageKeysEqual(storageKey, aStorageKey);
 }
+
+bool nsIGlobalObject::IsSystemPrincipal() const {
+  MOZ_ASSERT(NS_IsMainThread(),
+             "Cannot ask nsIGlobalObject IsSystemPrincipal off-main-thread");
+
+  return PrincipalOrNull()->IsSystemPrincipal();
+}
+
+RTPCallerType nsIGlobalObject::GetRTPCallerType() const {
+  if (IsSystemPrincipal()) {
+    return RTPCallerType::SystemPrincipal;
+  }
+
+  if (ShouldResistFingerprinting()) {
+    return RTPCallerType::ResistFingerprinting;
+  }
+
+  if (CrossOriginIsolated()) {
+    return RTPCallerType::CrossOriginIsolated;
+  }
+
+  return RTPCallerType::Normal;
+}